### PR TITLE
update Pioneer DDJ-SB2 mapping for Mixxx 2.1

### PR DESCRIPTION
**Pull request by Be-ing**
Update Pioneer DDJ-SB2 for new effects and looping UIs in Mixxx 2.1 and fix various other issues. New mappings for [autoloop pad layer](https://mixxx.org/wiki/doku.php/pioneer_ddj-sb2#pads_-_auto_loop_mode) and [effects](https://mixxx.org/wiki/doku.php/pioneer_ddj-sb2#effects) are documented on the wiki.

I did this while I was staying with a friend who happened to have a DDJ-SB2. I do not have access to the device to test further changes.

It would be nice to finish rewriting this using Components, but I did not have time to do that.

Depends on:
  - [x] #1227
  - [x] #1187
  - [x] #1271

### CODE REVIEW

**Comment by daschuer** [comment]
Who is able to review this?

**Comment by Be-ing** [comment]
@nikmartin? @DJMaxergy ? @dg3nec ?

**Comment by nikmartin** [comment]
Absolutely, just pulled out my SB2 and will test today. This issue states that it's dependent on #1127 and #1187. Should I merge those two PRs into master to test this? Since #1151 is now in master, do I need to do anything to make Tango use the new controls in PR #1187? I want to make sure I have my test environment setup correctly.

**Comment by Be-ing** [comment]
Both PRs this depended on have been merged to master. You can test it with the version of Tango in master, but you won't see the beatloop_size or beatjump_size on screen unless you use #940.

**Comment by Be-ing** [comment]
@nikmartin, can you report whether Mixxx loads the positions of the knobs and faders of the controller on startup? If so, please try commenting out lines 183 and 184 in the JS file, one line at a time, to test which one of those lines is working. Put `//` in front of the line to comment it out.

**Comment by Be-ing** [comment]
pawky2 on IRC tested this and discovered a nasty bug with the new mappings for the autoloop pads, which especially affected the 1st pad on the left. On button press, it should set a beatloop. On button press with shift, it should start a rolling beatloop, and on button up, it should exit the loop and exit slip mode to jump ahead to where the track would have been without the loop playing. The was some bizarre behavior after pressing it with shift depending on whether the pad or the shift button was released first. Initially, I commited afafe47 as a quick hack around the issue. However, I think similar issues with shift were affecting all the other pads in autoloop mode. It would have been ugly to implement similar hacks asafafe47 for every Button Component. I realized the right solution would be to change how the Components library handles Button input, hence PR #1271.

tl;dr for testers: You need to merge #1271 into this to test it for now.

**Comment by dec0de** [comment]
now, finally we have the possibility to turn all the knobs on the FX's by using shift+{trim,HI,MID,LO,Filter} knobs. The new auto loop functionality is really great once you get the hang of it. :-)

**Comment by nikmartin** [comment]
> can you report whether Mixxx loads the positions of the knobs and faders of the controller on startup? If so, please try commenting out lines 183 and 184

Th enumber may have changed by now, but this line is the one that works:
`midi.sendShortMsg(0x9B, 0x09, 0x7f);`

**Comment by sblaisot** [comment]
is that firmware version dependent ? in this case, we need to handle both or explain how to upgrade controller firmware version.

**Comment by Be-ing** [comment]
There don't seem to be any firmware updates available from Pioneer, so I think there only is one firmware version.

**Comment by Be-ing** [comment]
Thanks for confirming that @nikmartin. @dec0de found that to be the case as well. Commit c044b14 removed sending the unnecessary "DJ App Connect" message from Pioneer's MIDI documentation that does not seem to do anything. I think this is ready for merge as soon as #1271 is merged.

**Comment by nikmartin** [comment]
Great job so far. I need to read a bit for on the FX buttons, I seem to be able to get them in a weird state. The fx widget in mixxx has what looks like three states depending on the combination of pushes on the SB2's FX buttons, like Shift + FX, etc

**Comment by Be-ing** [comment]
I have written how the new effects mapping works on the [wiki page](https://mixxx.org/wiki/doku.php/pioneer_ddj-sb2#effects). Is there anything you're confused about after reading that? Which skin are you using?

**Comment by nikmartin** [comment]
@sblaisot I can't speak to being firmware version dependent, but the lower end Pioneer controllers almost never get a FW update. The one proprietary feature in these controllers is the PAd Trans effect, and If I could figure out how to do it in mixxx, I'd be in heaven. A very good effect. It's very similar to the new Loop Roll @Be-ing added to the SB2 javascript, but with a tremolo volume effect along with it. Like each time the loop starts, the volume quickly fades up and then back down as the loop ends

**Comment by Be-ing** [comment]
> The one proprietary feature in these controllers is the PAd Trans effect, and If I could figure out how to do it in mixxx, I'd be in heaven.

I cannot find much information about what this does. From [Pioneer's demo video](https://www.youtube.com/watch?v=OXbXgx7C-EA), it sounds cool. It sounds like a rolling loop combined with beat synchronized automation of a high pass filter, but I'm not sure if there is more to it than that. It would be cool to implement a generic way to automate effect parameters. Regardless, I don't think there is a way to do that with Mixxx's mapping system and I don't think controller mapping should be responsible for automating effects parameters.

**Comment by nikmartin** [comment]
> It sounds like a rolling loop combined with beat synchronized automation of a high pass filter, but I'm not sure if there is more to it than that.

I think what you are hearing as a high pass filter is actually a tremolo, where the volume quickly rises from 0 to [current channel-volume] and back in time with the track. the 4 buttons do 1/8, 1/4, 1/2 and 1 beat volume cuts. It's certainly doable in script, by getting the track bpm and the figuring out a setinterval time for each effect period. Now if it will perform is another question. I might mock it up to see if it's even feasible to do it in javascript.

**Comment by nikmartin** [comment]
@Be-ing 
> I have written how the new effects mapping works on the wiki page. Is there anything you're confused about after reading that? Which skin are you using?

After playing with it for a while I got the hang of it and it makes sense now.  With the Tango skin there appears to be a third state. I'll get a screenshot when I'm in front of my controller again.

**Comment by nikmartin** [comment]
I built a fresh copy of mixxx with #1271 , and the controller comes up in a weird state now. I'm using the Tango skin. I have samples that autoload into samplers 1-8. When mixxx starts, both deck hot cues buttons are lit, but deck 2's top row performance pads are also lit. This corresponds to hot cue 1-4. if i press one, it plays the sample, with the hot cue button lit. if i click the sampler button, it lights, the hot cue goes out, and the samplers still work. if i click the hot cue button again, the sampler light goes out, the hot cue light comes on, and the top row of performace pads goes out.

**Comment by Be-ing** [comment]
@dec0de also reported some weird issues on startup with the pad mode. Does this occur after replugging your controller then starting Mixxx? It might be that some of the MIDI messages being sent on shut down to turn the LEDs off are changing the state of the controller in a weird way. I added a [special case](https://github.com/mixxxdj/mixxx/pull/1243/files#diff-c440aa661d7afb2a0f89ba500aa462fbR372) to the shutdown code to avoid sending the message that would switch to decks 3 & 4 so those lights wouldn't remain on after shutdown. Perhaps that loop on shutdown is sending messages to change the pad mode. Sending a MIDI message in the `init` function to explicitly set the controller to specific pad modes might solve the issue. I'd presume that sending the controller back the same message it sends to the computer when one of the pad mode buttons is pressed switches it to that pad mode.

**Comment by Be-ing** [comment]
Try adding this to the `init` function:
```
for (var midiChannel = 1; midiChannel <= 4; midiChannel++) {
    midi.sendShortMsg(0x90 + midiChannel, 0x1B, 0x7F);
}
```
Does it switch the pads to hotcue mode on startup?

**Comment by nikmartin** [comment]
> Does this occur after replugging your controller then starting Mixxx?

The controller has remained plugged in. I also deleted the controller directory and then re-selected the controller in settings to make sure I was getting fresh files, etc. and the same thing happened. I always start mixxx with the controller plugged. I'll add your suggested code to the init function and report back.

**Comment by nikmartin** [comment]
Yes, I added that code to the end of init, and the controller came up in   'nominal' state. If I exit mixxx with a sampler bank active, the sampler button stays lit, but when I start mixxx again, everything resets and the hotcue buttons light up after mixxx loads.

**Comment by Be-ing** [comment]
Do the pads act like they're in hotcue mode on startup or does it just light the HOTCUE mode button?

**Comment by nikmartin** [comment]
No, the performance pads are not lit, just the hotcue button is. Since there's no tracks loaded, I wouldn't expect the pads to be lit. Pressing them doesn't do anything either.

**Comment by nikmartin** [comment]
interestingly, the loop buttons DO act like they are setting loops, even with no tracks loaded

**Comment by Be-ing** [comment]
Okay. When you load a track after starting Mixxx without doing anything else first, do the pads set/jump to hotcues?

**Comment by nikmartin** [comment]
I loaded a track that has hotcues immediately on startup, and the first three pads lit, corresponding to the three hotcues in the track. Also, the hotcues work properly right after loading.

**Comment by Be-ing** [comment]
Okay, so sending those MIDI messages does change the MIDI messages the controller sends. We could add an easy to configure user preference to pick the pad mode to set the controller to on startup. However, I think it would be better to avoid changing it on shutdown. In the `shutdown` function, try replacing:
```
    for (var channel = 0; channel <= 10; channel++) {
        for (var control = 0; control <= 127; control++) {
            // skip deck toggle buttons for now
            if (control === 0x72) {
                continue;
            }
            midi.sendShortMsg(0x90 + channel, control, 0);
        }
    }
```
with this:
```
    var skip = [0x72, 0x1B, 0x69, 0x1E, 0x6B, 0x20, 0x6D, 0x22, 0x6F, 0x70, 0x75];
    for (var channel = 0; channel <= 10; channel++) {
        for (var control = 0; control <= 127; control++) {
            // skip deck toggle buttons and pad mode buttons
            if (skip.indexOf(control) > -1) {
                continue;
            }
            midi.sendShortMsg(0x90 + channel, control, 0);
        }
    }
```

**Comment by nikmartin** [comment]
OK, now on shutdown, everything 'cleans up' and the controller is left in a nominal state.

**Comment by nikmartin** [comment]
Another not so small change is that shift+play used to reverse the track only while the play button was depressed. Now, it stays reversed until you release and hit play again. Can that be changed into an option in the script header? I use that function pretty heavily to censor tracks that I don't have a clean version of, and now it's not very easy to use when you are back masking over a single bad word.

**Comment by nikmartin** [comment]
I wrote a VERY close approximation of the native Pad Trans effect in the SB2. It's actually wired to the padtrans key in the midi xml, so the implementation below is hard coded to 1/8 of a beat. I think I may replace the Pads - kill (shift + manual loop) to see how it works in practice. If the pad trans wasn;t a proprietary non-midi control, we could use it as an on / off, but it takes over the pads midi values when it's on.
```
PioneerDDJSB2.ptMuted = [false,false];
PioneerDDJSB2.padTrans = function(channel, control, value, status, group){
    print('channel: ' + channel + ' control: ' + control + ' value: ' + value + ' status: ' + status + ' group: ' + group);
    var bpm = engine.getValue(group,'bpm');
    /**
    * transRate is how often to toggle the effect per BEAT (not second)
    * To calculate, take the bpm and divide by 60 (secs), to get how many beats per second
    * then, divide 1000 msec by that. and you get the interval in msec of one beat. The last calculation 
    * is to divide the interval by how many times per beat you want the effect. The stock SB2 does 1, 1/2, 1/4, and 1/8 of a beat
    * lets take a 128 bpm song for example:
    * 128 / 60 = 2.13 Beats per Second
    * 1000 / 2.13 = 468.75 is how many msec 1 beat is.
    */
    var transRate = (1000/(bpm/60))/16; // <<< hard coded to 1/8 beat right now
    if (value===0){
        engine.stopTimer(PioneerDDJSB2.pttimer);
        engine.setValue(group, 'mute', 0);
        return;
    }

    PioneerDDJSB2.pttimer = engine.beginTimer(transRate, function(){
              engine.setValue(group, 'mute', PioneerDDJSB2.ptMuted[channel]); 
             PioneerDDJSB2.ptMuted[channel] = !PioneerDDJSB2.ptMuted[channel];   
    }, false);
};
```

**Comment by Be-ing** [comment]
> OK, now on shutdown, everything 'cleans up' and the controller is left in a nominal state.

When you start Mixxx again, is the pad mode the same as when Mixxx shut down?

> Another not so small change is that shift+play used to reverse the track only while the play button was depressed. Now, it stays reversed until you release and hit play again. Can that be changed into an option in the script header? I use that function pretty heavily to censor tracks that I don't have a clean version of, and now it's not very easy to use when you are back masking over a single bad word.

Hmm, could that be fit into the mapping elsewhere? Maybe shift + Pad Trans? That could get tricky though. I think the mapping would have to track the active pad mode and reset it to the last mode when shift + Pad Trans was released.

> I wrote a VERY close approximation of the native Pad Trans effect in the SB2. It's actually wired to the padtrans key in the midi xml, so the implementation below is hard coded to 1/8 of a beat. I think I may replace the Pads - kill (shift + manual loop) to see how it works in practice. If the pad trans wasn;t a proprietary non-midi control, we could use it as an on / off, but it takes over the pads midi values when it's on.

Proprietary non-MIDI control? It's a USB MIDI class compliant device. Do you mean the pads send different MIDI messages when Pad Trans is active? That's how the other pad modes work. I don't see those MIDI codes listed in [Pioneer's documentation](https://pdj-ecom-cdn.azureedge.net/-/media/pioneerdj/software-info/controller/ddj-sb2/ddj-sb2_list_of_midi_messages_e.pdf?la=en) though, so it would be up to you to map those. You could make a pull request for my ddj_sb2_mapping branch, or we could merge this, then you could make another pull request for the master branch of this repository.

**Comment by nikmartin** [comment]
> When you start Mixxx again, is the pad mode the same as when Mixxx shut down?

Upon further debugging, on shutdown, almost everything is reset to nominal with just the hotcue button on, and it comes up that same way. It does reset the vinyl mode, but that's a javascript variable. If the Shift + Manual Loop light is blinking, it remains that way when mixxx exits, and when restarted, everything is set back to a nominal state.

> Proprietary non-MIDI control? It's a USB MIDI class compliant device.

It is, except for the pad-trans button. it sends  

`Debug [Controller]: "DDJ-SB2 MIDI 1: t:4008244 ms status 0x90 (ch 1, opcode 0x9), ctrl 0x70, val 0x00"` 

On channel 1, BUT it somehow short circuits the performance pads 1-4. While the padTrans button is enabled, pads 1-4 (top row ) all send:

`Debug [Controller]: "DDJ-SB2 MIDI 1: t:4140176 ms status 0xB0 (ch 1, opcode 0xB), ctrl 0x13, val 0x27"`
`Debug [Controller]: "DDJ-SB2 MIDI 1: t:4140176 ms status 0xB0 (ch 1, opcode 0xB), ctrl 0x33, val 0x00"`

And only send that when they are released. There's a lot of grumbling about this over at Traktor. In fact, if you start Serato once, use the pad trans effects, then exit and start mixxx, the pad trans effect works until you unplug the controller. The effect is actually loaded in the controller until it's unplugged.

**Comment by Be-ing** [comment]
Hmm, that's strange. Maybe when the Pad Trans button is pressed Serato sends the controller a MIDI message that sets the controller to a state where it sends distinct messages for each pad. I doubt they went so far out of their way to do something that isn't a standard MIDI message. Perhaps that mysterious "DJ App Connect" message has something to do with it. If you want to get that to work, I suggest using [MIDI Monitor](http://www.snoize.com/MIDIMonitor/) (Mac) or [MIDI OX](http://www.midiox.com/) (Windows) to intercept the MIDI messages that Serato sends to the controller.

**Comment by nikmartin** [comment]
Yes, my guess is that since the effect works after being disconnected from Serato and connected to mixxx, that the effect is actually being done in the soundcard on the SB2, and the performance pads are no longer sending valid midi messages when the PAd Trans button is on.. You SB2 mapping is quite well written, but I'm having a time finding an unused latching button (like the Shift+Manual Loop button) that will let me add this effect to the mapping. I know that  the Filter Kill effect commands are now deprecated in lieu of the Effects framework, but I don't want to get rid of them until the commands are actually removed from mixxx. Any other ideas for an unused Shift + Button combo?

**Comment by Be-ing** [comment]
> the effect is actually being done in the soundcard on the SB2

I think it's highly unlikely that Pioneer put audio DSP inside the DDJ-SB2. That would be a huge waste of money for them. It doesn't even have an analog-to-digital converter for the microphone input. From [this video](https://www.youtube.com/watch?v=X-7aA4I_nE4), it looks like the firmware makes the effect just by sending messages as if the volume fader was jumping up/down really fast, like how the filter fade feature sends MIDI messages as if the filter knobs are being turned.

**Comment by nikmartin** [comment]
Agreed, the effect itself is pretty simple, so that's probably it. My implementation works pretty well, but the timer api _will_ drift out of sync after a while, but longer than you'd actually use the effect. My initial thought was to try to do it using the mixxx effects chain, assigning the effect to a high numbered effects slot, but couldn't grok the effects framework to figure it out.

**Comment by nikmartin** [comment]
> You could make a pull request for my ddj_sb2_mapping branch, or we could merge this, then you could make another pull request for the master branch of this repository.

Let me debug it a bit more and make it deck independent and find an unused effect enable key in the mapping, and I'll do a pull request

**Comment by Be-ing** [comment]
> While the padTrans button is enabled, pads 1-4 (top row ) all send:
```
Debug [Controller]: "DDJ-SB2 MIDI 1: t:4140176 ms status 0xB0 (ch 1, opcode 0xB), ctrl 0x13, val 0x27"
Debug [Controller]: "DDJ-SB2 MIDI 1: t:4140176 ms status 0xB0 (ch 1, opcode 0xB), ctrl 0x33, val 0x00"
```

Those are the MIDI messages for the deck volume fader (it uses 2 MIDI messages, one for a least significant byte and another for a most significant byte).

**Comment by Be-ing** [comment]
> Let me debug it a bit more and make it deck independent and find an unused effect enable key in the mapping, and I'll do a pull request

I think it would be a better idea to figure out how Serato communicates the tempo to the controller and work with how the controller is designed.

**Comment by sblaisot** [comment]
midi capture here : https://forums.pioneerdj.com/hc/en-us/community/posts/205846443-DDJ-SB2

**Comment by Be-ing** [comment]
@dec0de can you confirm that the issue with the pad mode on startup has been fixed with the latest commit?

**Comment by Be-ing** [comment]
@nikmartin: I added an option to use shift + cue for reverseroll like it was mapped in Mixxx 2.0.

Are there any issues remaining? Ready for merge? If someone who has the controller wants to implement the Pad Trans functionality, I think that should be done in a new PR after merging this.

File: res/controllers/Pioneer-DDJ-SB2-scripts.js [hunk 1]
```diff
@@ -10,6 +10,11 @@ PioneerDDJSB2.vinylModeOnStartup = false;
 // Allows toggling slip faster, but is counterintuitive.
 PioneerDDJSB2.invertVinylSlipButton = false;
 
+// If true, pressing shift + cue will play the track in reverse and enable slip mode,
+// which can be used like a censor effect. If false, pressing shift + cue jumps to
+// the beginning of the track and stops playback.
+PioneerDDJSB2.reverseRollOnShiftCue = false;
```
Comment (nikmartin) [inline res/controllers/Pioneer-DDJ-SB2-scripts.js]: Is this feature a temporary reverse roll, such that it resumes playing (slip mode) when the buttons are released?, or does it continue to play backwards until pressed again?

**Comment by nikmartin** [comment]
@Be-ing I will test heavily the next few days, and open new issues that I find. I will also make an attempt at the Pad Trans effect and do a PR based on master after this is merged. Based on someone's earlier reply to my comment about what the controller is sending on pad trans + effect buttons, the effect seems to rely on the controller knowing the BPM of the song, and the rapid volume switches are in sync with the bpm. My attempt is just using the bpm from the track, but is not synced in any way. I'm not sure that that is available to javascript from the mixxx engine. I don't think the controller has any DSP capability, but does seem to possibly know how to do beat detection? Not sure yet.

**Comment by Be-ing** [comment]
@daschuer: could you merge this?

@nikmartin: I highly doubt the controller has any way to detect the beat. It probably gets that information from Serato with some undocumented MIDI message. You'd have to intercept the MIDI messages from Serato to find that. Mixxx does make the BPM available to controller mappings through the "bpm" Control.

**Comment by daschuer** [comment]
Thank you for all the work. And for reviewing this.
